### PR TITLE
Define det(Q) for {QR,LQ}PackedQ and QRCompactWYQ

### PR DESCRIPTION
**Pull request by tkf**
This PR adds a one-linear `det(Q::Union{QRPackedQ, LQPackedQ}) = isodd(count(!iszero, Q.τ)) ? -1 : 1` (and a few tests; not sure if they are useful though).  It counts the number of Householder reflections and returns `1` or `-1` depending on the parity.

Does it make sense to have it in stdlib?  Also, is this definition valid?  I feel a bit uneasy to do `iszero` on floats.

### CODE REVIEW

**Comment by andreasnoack** [comment]
I'd have to think a little more about this but at least the non-real cases are not covered by your definition.

**Comment by tkf** [comment]
I realized that my definition assumes `τ` to be 0 or 2 (as `τ = 2` was the definition I found in Wikipedia).  But it's easy to get the determinant of the reflector

```
H = I - τ v v'
```

for arbitrary `τ` since:

* For any `u` orthogonal to `v`, we have `H u = u`.  This gives us `size(H, 1)` eigenvalues of 1.
* There is one more eigenvalue `(1 - τ)` since `H v = (1 - τ) v`.

So, it's `det(H) = 1 - τ`.  The correct implementation would be

```julia
det(Q::Union{QRPackedQ, LQPackedQ}) = prod(1 .- Q.τ)
```

> at least the non-real cases are not covered by your definition.

I don't quite get this. Isn't `QRPackedQ`/`LQPackedQ` using the same definition of `τ` for complex values?  Is the new definition resolves your concern?

**Comment by andreasnoack** [comment]
>  Is the new definition resolves your concern?

Yes, your new definition applies generally. However, there is a complication due to the way LAPACK stores the reflectors. To reduce the space required to store the vectors, they are scaled by their first element and the scaling is included in `τ`. This makes the computation a bit more complicated and also a bit more expensive. For each reflector, the non-unit eigenvalue is `1 - v'v*τ` in the LAPACK storage format which is therefore more like `1 - (1 + v[2:end]'v[2:end])*τ` because the first element is implicit.

**Comment by tkf** [comment]
I see. I missed that LAPACK manual doesn't say that `v` is normalized. Thanks for the explanation. I'll incorporate `v` in the computation.

File: stdlib/LinearAlgebra/src/qr.jl [hunk 1]
```diff
@@ -906,3 +906,14 @@ end
...
+
+# Compute `det` from the number of Householder reflections.
+det(Q::QRPackedQ) = _det_qr(Q.factors, Q.τ)
+_det_qr(factors, τ) =
+    prod(axes(τ, 1)) do i
+        Base.@_inline_meta
+        # "v" in the math definition except for the diagonal (= 1):
+        v = @inbounds @view factors[i + 1:end, i]
+        𝟙 = one(eltype(factors))
+        𝟙 - (𝟙 + v'v) * @inbounds τ[i]
```
Comment (tkf) [inline stdlib/LinearAlgebra/src/qr.jl]: FYI, it's much faster (~ 1.5x) to directly implement it using `for` loop as below for moderate-size matrices (~ 100).  The difference is bigger (~ 2.4x) without `@inline` annotation added in 940d16c85d5231a6bd3235f789fcef8968bf266f.

```julia
_det_qr(factors, τ) =
    prod(axes(τ, 1)) do i
        Base.@_inline_meta
        vv = 𝟙 = one(eltype(factors))
        @simd for j in axes(factors, 1)[i + 1:end]
            vv += abs2(@inbounds factors[j, i])
        end
        𝟙 - vv * @inbounds τ[i]
    end
```

But I thought it'd be much better to tweak `v'v` in such a way that the code like above is unnecessary.

**Comment by tkf** [comment]
I incorporated `Q.factors` in the computation of `det(Q)`.

File: stdlib/LinearAlgebra/src/qr.jl [hunk 2]
```diff
@@ -906,3 +906,14 @@ end
...
 # FIXME! Should add balancing option through xgebal
+
+# Compute `det` from the number of Householder reflections.
+det(Q::QRPackedQ) = _det_qr(Q.factors, Q.τ)
+_det_qr(factors, τ) =
+    prod(axes(τ, 1)) do i
+        Base.@_inline_meta
+        # "v" in the math definition except for the diagonal (= 1):
+        v = @inbounds @view factors[i + 1:end, i]
+        𝟙 = one(eltype(factors))
```
Comment (andreasnoack) [inline stdlib/LinearAlgebra/src/qr.jl]: Did you try with `1`? It usually promotes correctly in float computations and I don't think there end up being any overhead.

File: stdlib/LinearAlgebra/src/qr.jl (hunk 1 above)
Comment (andreasnoack) [reply stdlib/LinearAlgebra/src/qr.jl]: I agree that the current version looks better

File: stdlib/LinearAlgebra/src/matmul.jl [hunk 3]
```diff
@@ -6,8 +6,8 @@ matprod(x, y) = x*y + x*y
 
 # dot products
 
-dot(x::Union{DenseArray{T},StridedVector{T}}, y::Union{DenseArray{T},StridedVector{T}}) where {T<:BlasReal} = BLAS.dot(x, y)
-dot(x::Union{DenseArray{T},StridedVector{T}}, y::Union{DenseArray{T},StridedVector{T}}) where {T<:BlasComplex} = BLAS.dotc(x, y)
+@inline dot(x::Union{DenseArray{T},StridedVector{T}}, y::Union{DenseArray{T},StridedVector{T}}) where {T<:BlasReal} = BLAS.dot(x, y)
```
Comment (andreasnoack) [inline stdlib/LinearAlgebra/src/matmul.jl]: Does inlining of the BLAS versions really matter?

**Review by andreasnoack — COMMENTED** [review]
Looks good. I wondered if we could somehow avoid accessing the reflector vectors by exploiting the fact that the eigenvalue is on the unit circle. I.e. we know that `abs2(1 - c*τ[i]) == 1` for `c>0`. We could just solve the equation but that would require computing square roots. For larger matrices that might still be cheaper than accessing all the reflectors elements. Can we do better? We know we'll end up on the unit circle.

**Comment by tkf** [comment]
> I.e. we know that `abs2(1 - c*τ[i]) == 1` for `c>0`

Interesting. So, we have (omitting `[i]`):

```julia
abs2(1 - c*τ)
== (1 - c*τ) * (1 - conj(c*τ))
== 1 + abs2(c*τ) - 2real(c*τ)
== 1 + c * (c * abs2(τ) - 2real(τ))
```

which indicates that `c = 2real(τ) / abs2(τ)` works if `real(τ) > 0`.  Is it guaranteed?  I know that `τ` can be zero but if we know `real(τ) >= 0` then I guess we can avoid square root.  I guess it makes sense as `real(τ) < 0` doesn't feel like "reflecting" the vector.  Falling back to square root if `real(τ) < 0` may not be so bad, though.

But, with this approach, how should we handle very small `abs2(τ)` with `!iszero(τ)`? (Is it even possible?)  Maybe falling back to `1 - v'v*τ` when `0 < abs2(τ) < small_tau_threshold`?  What `small_tau_threshold` should we use?

**Comment by andreasnoack** [comment]
Great. That must be the way to do it. Took a look and the eigenvalue actually simplifies to `-sign(conj(τ))^2` so there shouldn't be any need for special handling of small values. It is probably also a good idea to add a method for the real case similar to your original version (I've now convinced myself that it must be correct).

**Comment by tkf** [comment]
> simplifies to `-sign(conj(τ))^2`

Great! I updated the PR.

**Comment by tkf** [comment]
It looks like we need to handle `iszero(τ)` case explicitly (7516fd0276c3167fdbdf98a23b06a5e9f8c6ef0c) because it can happen in complex matrices:

```julia
julia> qr(cat(Matrix(I, 3, 3), randn(ComplexF64, 2, 2), dims=(1, 2)), Val(true)).τ
5-element Array{Complex{Float64},1}:
                1.0 + 0.0im
                0.0 + 0.0im
                0.0 + 0.0im
 1.5403413193290247 + 0.11524192317353117im
 1.2489469327380043 + 0.9685171266840557im
```

Also, it looks like the eigenvalue should be `-sign(τ)^2` (no `conj`).

File: stdlib/LinearAlgebra/src/lq.jl [hunk 4]
```diff
@@ -333,3 +333,25 @@ function ldiv!(A::LQ{T}, B::StridedVecOrMat{T}) where T
     lmul!(adjoint(A.Q), ldiv!(LowerTriangular(A.L),B))
     return B
 end
+
+
+_tau(Q) = Q.τ
+_tau(Q::QRCompactWYQ) = @view Q.T[diagind(Q.T)]
```
Comment (andreasnoack) [inline stdlib/LinearAlgebra/src/lq.jl]: Did you do the math for this case? Great if it also works for blocks.

File: stdlib/LinearAlgebra/src/lq.jl [hunk 5]
```diff
@@ -333,3 +333,25 @@ function ldiv!(A::LQ{T}, B::StridedVecOrMat{T}) where T
...
+
+# Compute `det` from the number of Householder reflections.  Handle
+# the case `Q.τ` contains zeros.
+det(Q::Union{QRPackedQ{<:Real}, QRCompactWYQ{<:Real}, LQPackedQ{<:Real}}) =
+    isodd(count(!iszero, _tau(Q))) ? -1 : 1
```
Comment (andreasnoack) [inline stdlib/LinearAlgebra/src/lq.jl]: Maybe we should use the element type as the return type here

File: stdlib/LinearAlgebra/src/lq.jl (hunk 4 above)
Comment (tkf) [reply stdlib/LinearAlgebra/src/lq.jl]: Reading the docs https://docs.julialang.org/en/latest/stdlib/LinearAlgebra/#LinearAlgebra.QRCompactWY, I thought that I can just use the diagonal to extract the tau.  But that's actually not the case when `minimum(size(A)) > 36`?

According to LAPACK docs http://www.netlib.org/lapack/explore-html/dd/d9a/group__double_g_ecomputational_gaddcf152e87deec6123a1899f6f51101e.html#gaddcf152e87deec6123a1899f6f51101e, `F.T` is an `hcat`ed upper triangular matrices `T_n`, i.e.,

```
F.T == [T_1 T_2 ... T_B]
```

Does it mean

```
tau == [diag(T_1); diag(T_2); ...; diag(T_B)]
```

?

**Comment by tkf** [comment]
@andreasnoack Can this be merged?

Regarding blocked QR (https://github.com/JuliaLang/julia/pull/32887#discussion_r315559951), now that I documented the storage format of QRCompactWY in https://github.com/JuliaLang/julia/pull/32996, I'm more confident about the implementation I added in 7b8a2d3f81c80d8a91b0bd15f52e1c58aa5e9d9b.  But I understand if you need more time to review math and/or implementation.

**Comment by andreasnoack** [comment]
Yes. I think this should be good to go.

**Pull request merged**